### PR TITLE
[win32] Fix trac #16165

### PR DESCRIPTION
**Pull request by afedchin**
This fixes http://trac.kodi.tv/ticket/16165 and partially reverts #2500

@Paxxi @topfs2 please take a look.

### CODE REVIEW

**Comment by hudokkow** [comment]
jenkins build this please

File: xbmc/powermanagement/windows/Win32PowerSyscall.cpp [hunk 1]
```diff
@@ -25,35 +25,67 @@
...
+  }
+
+  return false;
+}
+
+void CWin32PowerStateWorker::Process(void)
+{
+  while (!m_bStop)
+  {
+    if (m_queryEvent.WaitMSec(100))
```
Comment (Paxxi) [inline xbmc/powermanagement/windows/Win32PowerSyscall.cpp]: I don't see a need to keep waking up here, at least not this often?

File: xbmc/powermanagement/windows/Win32PowerSyscall.h [hunk 2]
```diff
@@ -24,11 +24,32 @@
...
+  ~CWin32PowerStateWorker() { m_queryEvent.Reset(); }
+  bool QueryStateChange(PowerState State);
+
+protected:
+  virtual void Process(void);
+  virtual void OnStartup() { SetPriority(THREAD_PRIORITY_IDLE); };
+
+private:
+  static CCriticalSection m_critSection;
+  PowerState m_state;
```
Comment (Paxxi) [inline xbmc/powermanagement/windows/Win32PowerSyscall.h]: this can be an std::atomic< PowerState > and then the lock isn't needed.

**Comment by Paxxi** [comment]
Not runtime tested but the logic seems sound.

**Comment by afedchin** [comment]
updated.

**Comment by afedchin** [comment]
jenkins build this please

**Comment by afedchin** [comment]
@MartijnKaijser should this be backported?

**Comment by MartijnKaijser** [comment]
don't know tbh. Depends what others think

**Comment by mkortstiege** [comment]
A bit late here, but isn't removing the `PumpPowerEvents` killing the `OnSleep`and `OnWake` callbacks like eg. update library on wake?

**Comment by afedchin** [comment]
@mkortstiege It's never late. I haven't understood you. The PumpPowerEvents is required to process OnSuspend/OnResume events.

File: xbmc/powermanagement/windows/Win32PowerSyscall.cpp [hunk 3]
```diff
@@ -26,34 +26,59 @@
...
 bool CWin32PowerSyscall::Powerdown()
 {
-  return CWIN32Util::PowerManagement(POWERSTATE_SHUTDOWN);
+  return m_worker.QueryStateChange(POWERSTATE_SHUTDOWN);
 }
 bool CWin32PowerSyscall::Suspend()
 {
-  // On Vista+, we don't receive the PBT_APMSUSPEND message as we have fired the suspend mode
-  // Set the flag manually
-  CWin32PowerSyscall::SetOnSuspend();
```
Comment (Paxxi) [inline xbmc/powermanagement/windows/Win32PowerSyscall.cpp]: This call should still be here, it sets a flag that is then read in processevents deciding which callback to call

File: xbmc/powermanagement/windows/Win32PowerSyscall.cpp [hunk 4]
```diff
@@ -26,34 +26,59 @@
...
-
-  return CWIN32Util::PowerManagement(POWERSTATE_SUSPEND);
+  return m_worker.QueryStateChange(POWERSTATE_SUSPEND);
 }
 bool CWin32PowerSyscall::Hibernate()
 {
-  // On Vista+, we don't receive the PBT_APMSUSPEND message as we have fired the suspend mode
-  // Set the flag manually
-  CWin32PowerSyscall::SetOnSuspend();
```
Comment (Paxxi) [inline xbmc/powermanagement/windows/Win32PowerSyscall.cpp]: same here

File: xbmc/win32/WIN32Util.cpp [hunk 5]
```diff
@@ -216,18 +216,15 @@ bool CWIN32Util::PowerManagement(PowerState State)
       return false;
   }
 
-  // process OnSleep() events. This is called in main thread.
-  g_powerManager.ProcessEvents();
```
Comment (Paxxi) [inline xbmc/win32/WIN32Util.cpp]: I think @mkortstiege is correct that we still need this, it doesn't process anything but it calls the callbacks on g_powerManager to notify about suspend/resume events

**Comment by Paxxi** [comment]
Ignore my comments, seems I was wrong and everything should be fine

**Comment by afedchin** [comment]
jenkins build this please

**Comment by afedchin** [comment]
@Paxxi @mkortstiege @MartijnKaijser if no objections I'll merge this in next two days.

**Comment by mkortstiege** [comment]
No objections from my side.

**Comment by MartijnKaijser** [comment]
backport for 15.2 ?

**Comment by afedchin** [comment]
will do today

**Pull request merged**